storages/jsonl-storage/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+                    })
+                    .unwrap_or(false)
+            })
+            .map(|result| -> Result<_> {
+                let path = result.map_storage_err()?.path();
+                let table_name = path
+                    .file_stem()
+                    .map_storage_err(JsonlStorageError::FileNotFound.to_string())?
+                    .to_str()
+                    .map_storage_err(JsonlStorageError::FileNotFound.to_string())?
+                    .to_owned();
+
+                self.fetch_schema(table_name.as_str())?
+                    .map_storage_err(JsonlStorageError::TableDoesNotExist.to_string())
+            })
+            .collect::<Result<Vec<Schema>>>()?;
+
+        schemas.sort_by(|a, b| a.table_name.cmp(&b.table_name));
+
+        Ok(schemas)
+    }
+
+    async fn fetch_data(&self, table_name: &str, target: &Key) -> Result<Option<DataRow>> {
+        let row = self.scan_data(table_name)?.find_map(|result| {
+            result
+                .map(|(key, row)| (&key == target).then_some(row))
+                .unwrap_or(None)
+        });
+
+        Ok(row)
+    }
+
+    async fn scan_data(&self, table_name: &str) -> Result<RowIter> {
+        self.scan_data(table_name)
+    }
+}
+
+fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
+where
+    P: AsRef<Path>,
+{
+    let file = File::open(filename)?;
+    Ok(io::BufReader::new(file).lines())
+}
+
+type IsLeft = bool;
+struct SortMerge<T: Iterator<Item = Result<(Key, DataRow)>>> {
+    prev_rows: T,
+    rows: IntoIter<(Key, DataRow)>,
+    current: Option<(IsLeft, Key, DataRow)>,
+}
+
+impl<T> SortMerge<T>
+where
+    T: Iterator<Item = Result<(Key, DataRow)>>,
+{
+    fn new(prev_rows: T, rows: IntoIter<(Key, DataRow)>) -> Self {
+        Self {
+            prev_rows,
+            rows,
+            current: None,
+        }
+    }
+}
+
+impl<T: Iterator<Item = Result<(Key, DataRow)>>> Iterator for SortMerge<T> {
+    type Item = Result<DataRow>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut result = || -> Result<_> {
+            let left_right = match self.current.take() {
+                Some((true, key, row)) => (Some((key, row)), self.rows.next()),
+                Some((false, key, row)) => (self.prev_rows.next().transpose()?, Some((key, row))),
+                None => (self.prev_rows.next().transpose()?, self.rows.next()),
+            };
+
+            Ok(left_right)
+        };
+
+        let (left, right) = match result() {
+            Ok(left_right) => left_right,
+            Err(e) => return Some(Err(e)),
+        };
+
+        let data_row = match (left, right) {
+            (None, None) => None,
+            (None, Some((_, row))) => {
+                self.current = None;
+
+                Some(row)
+            }
+            (Some((_, prev_row)), None) => {
+                self.current = None;
+
+                Some(prev_row)
+            }
```

Comment:
Instead of these,
```rust
(left, right) => {
    self.current = None;

    left.or(right).map(|(_, row)| row)
}
```
You can put this after the `Some(_), Some(_)`.